ocrd_webapi/authentication.py

```diff
@@ -0,0 +1,52 @@
+from hashlib import sha512
+from random import random
+from typing import Tuple
+
+from .database import create_user, get_user
+
+
+async def authenticate_user(email: str, password: str):
+    db_user = await get_user(email=email)
+    if not db_user:
+        raise ValueError(f"User not found: {email}")
+    validation_status = validate_password(
+        plain_password=password,
+        encrypted_password=db_user.encrypted_pass
+    )
+    if not validation_status:
+        raise ValueError(f"User was not validated: {email}")
+
+
+async def register_user(email: str, password: str):
+    salt, encrypted_password = encrypt_password(password)
+    db_user = await get_user(email)
+    if db_user:
+        raise ValueError(f"User is already registered: {email}")
+    created_user = await create_user(
+        email=email,
+        encrypted_pass=encrypted_password,
+        salt=salt,
+        validated_account=False
+    )
+    if not created_user:
+        raise ValueError(f"Failed to register user: {email}")
+
+
+def encrypt_password(plain_password: str) -> Tuple[str, str]:
+    salt = get_random_salt()
+    hashed_password = get_hex_digest(salt, plain_password)
+    encrypted_password = f'{salt}${hashed_password}'
+    return salt, encrypted_password
+
+
+def get_hex_digest(salt: str, plain_password: str):
+    return sha512(f'{salt}{plain_password}'.encode('utf-8')).hexdigest()
+
+
+def get_random_salt() -> str:
+    return sha512(f'{str(random())}{hash}'.encode('utf-8')).hexdigest()[:8]
```

Comment:
I am not sure what was intended here, but I think the `{hash}` is somehow wrong here. Maybe: `f'{hash(str(random()))}` ?

Reply:
The way it is now is not problematic because the hash here does not have to be the hash of the `str(random())` specifically. It's instead an additional randomization value. Ultimately, the `salt` value is always 8 chars long HEX.

Reply:
Look at what the code does (`{str(random())}{hash}`). Currently it is something like this: `0.6931815889940808<built-in function hash>` . It is not a problem but I don't think that it was intended this way. Why keep hard to understand code that does not what was intended? Or I got it wrong and you it provides something that I do not get, then please explain but I think it is just a mistake.

Reply:
> It is not a problem but I don't think that it was intended this way

> Why keep hard to understand code that does not what was intended? 

I will change it if it's confusing and seems to be for you. As I previously wrote, the idea is that we only care to get a hex digest of some random bytes of a specific length (8 in this case). So it does not matter that the `hash` does not get an object as a parameter. It just puts the memory pointer of the `hash` method in the byte mix. It was something I intended to make that way.